### PR TITLE
Destroyable lockers

### PR DESCRIPTION
**Pull request by flappybatpal**
[FEATURE][BALANCE]

- [x] Fix logging
- [x] Mostly resolve horrible code duplication (kinda hard due to types)
- [x] Make reinforced protect against weak kinetic but not strong?

## About the PR 
- Lockers with card locks can be beaten open with items that do at least 1 damage. When opened the door is broken and needs to be welded to close.
  - Command lockers have 300 health.
  - Lockers with card locks have 200 health.
  - Regular lockers have 100 health - Cannot be hit with items as they have no lock, just open it!
  - Items that do 10 damage or less are not very effective and do 70% damage.
  - Damage from items is capped to 20 to prevent high damage melee making them a joke (looking at you csaber).
- All lockers can be shot open with projectiles.
  - Kinetic weapons do 80% damage
  - Energy projectiles do 40% damage as lasers/phasers are easily available and rechargeable.
- Some special lockers are reinforced. This is currently the HoS, Armory and Confiscated Items locker.
  - Reinforced lockers cannot be damaged with items.
  - Energy projectiles do not damage reinforced lockers.
  - 26+ damage projectiles are required to break reinforced lockers. No cheap ammo like 9mm or .22.
  - Rubber slug does no damage to reinforced lockers to prevent you just using the NARCS turret.
- All card secured lockers start bolted to the floor and must be unbolted with a wrench to be moved.
- Unique sounds for damage and breaking

Uses the sounds:
https://freesound.org/people/skinnytecboy/sounds/188063/ used under [[CC0]](https://creativecommons.org/publicdomain/zero/1.0/)
https://freesound.org/people/SamsterBirdies/sounds/587443/ used under [[CC0]](https://creativecommons.org/publicdomain/zero/1.0/)
https://freesound.org/people/chripei/sounds/393659/ used under [[CC0]](https://creativecommons.org/publicdomain/zero/1.0/)

https://user-images.githubusercontent.com/70909958/162621254-4e0d82d9-9366-47cf-a963-8808f708538d.mp4

## Why's this needed? 
Lockers are strangely invincible except to explosions. While this sounds like strong protection the reality is they can easily be broken into with a rigged lightbulb or welding fuel pipebomb.

Lockers being largely indestructible made them the go-to against turrets and could present some frustrations to antagonists that breaking into a locker had both a slight knowledge and equipment requirement 

Breaking a locker busts it open. It cannot be closed until a welding tool is used on it.
Nukies/turrets will no longer be totally defeated by lockers. Energy weapons are less effective to reflect being quite common for non traitors.

The HoS and armory locker have special protection from their reinforced status to reflect they will contain an egun or requisition tokens. You have to expend lethal kinetic ammunition to break these making it a trade off.

## Changelog

```changelog
(u)Flappybat
(*)Closets and lockers can now be broken open by hitting them with items or shooting with projectiles. Lockers have between 100HP and 300HP depending on their security. Energy weapons and low damage items are less effective against them.
(*)HoS, Armory and Confiscated item lockers are reinforced and can only be broken open with explosives and higher damage kinetics.
(*)Secure lockers started bolted to the floor. They must be unbolted with a wrench to be moved.
```

### CODE REVIEW

**Comment by flappybatpal** [comment]
Changed health to flip_health to better describe what it does.
Changed some use of W for item to I.

I'm sure the code could be better. Didn't want to add the damage handling to the parent as crates belong to it and would inherit. Closets and secure lockers are from different types so they had to have repeat logic rather than inheriting.

There are a whole bunch of special interactions with attackby, think I have the right solution to not break the existing ones while avoiding players casually smashing into lockers because they are holding an item.

**Comment by ThatFiveGuys** [comment]
This could probably go with a forum thread. My main concern is the methods currently require preperation, whereas this you can easily do on a whim. I feel like this would make breaking into lockers much more common and dont know if thats a good thing.

**Comment by NightmareChamillian** [comment]
Suggestion: 
Any weapon below 10 damage does two thirds of it's damage to the locker.
This way, you can still bash them with really common items like extinguishers and crowbars, but in order to actually bust one with efficiency, you'll have to be pre-prepared with a decent melee item

File: code/obj/storage/secure_closets.dm [hunk 1]
```diff
@@ -4,6 +4,95 @@
...
+		if (!P || !istype(P.proj_data,/datum/projectile/))
+			return
+		if (reinforced)
+			// To prevent parking in front of NARCS for ammo free locker entry
+			if (istype(P.proj_data, /datum/projectile/bullet/abg))
+				return
+		damage = round((P.power*P.proj_data.ks_ratio), 1.0)
+		// .22 is quite high raw damage and can be manufactured. Custom is material dipped .22. Stinger grenades did 60 damage.
+		if (istype(P.proj_data, /datum/projectile/bullet/bullet_22) || istype(P.proj_data, /datum/projectile/bullet/custom) || istype(P.proj_data, /datum/projectile/bullet/stinger_ball))
+			damage = round(damage * 0.5)
```
Comment (Tarmunora) [inline code/obj/storage/secure_closets.dm:63]: typechecking on projectiles feels really weird and I don't like it

**Review by Tarmunora — CHANGES_REQUESTED** [review]
Agreed that some level of flat damage reduction wouldn't be a bad thing

File: code/obj/storage/secure_closets.dm (hunk 1 above)
Comment (flappybatpal) [reply code/obj/storage/secure_closets.dm:63]: Would a var be preferable?
Calibre has the issue of being shared.

Comment (Tarmunora) [reply code/obj/storage/secure_closets.dm:63]: could do a check on projectile power - all 3 of those projectiles are <25 power, for example

**Comment by flappybatpal** [comment]
I think this is feature complete, pending feedback.

**Comment by NightmareChamillian** [comment]
Would love to see a tiny bit of animation on the breaking and hit by projectile events.

**Comment by flappybatpal** [comment]
I thought about it but those can happen really rapidly.

**Comment by frawhst** [comment]
Gonna wait on some testing; sounds good mechanically, but I want to get a feel by running around and shooting some lockers. Will probably be a few days for me personally, but I can merge into develop later today for general testing.

**Review by virvatuli-gh — APPROVED** [review]
Already mentioned it in discord but the sounds are good. Only thing I could say is that `locker_break.ogg`'s attack is a little too sudden and loud, and the whole file could be a smidge shorter without the extra debris clanging at the end (for instances where the sound might be played multiple times rapidly), but that's not dealbreaking or anything. I'd rather hear it in a busy environment first anyways. Cool feature, no comment on balance from me though.

**Comment by frawhst** [comment]
From (very) minor testing-
- would be nice if secure lockers had *some* feedback to say 'yeah you're doing damage to this'. Could make them do what doors do when they get damaged where they spark a bit. Some kind of bullet impact sound would also be nice but I'm not sure we have those!!!
- Projectiles feel a bit strong; took about 1 orion mag or 5 revolver shots to bust the cap's locker. Didn't test with crew-accessible weapons.

**Comment by flappybatpal** [comment]
Reduced kinetic to 80%, energy to 40%.

I tried elec flash, it felt out of place particularly with the sound.
I have added a hit effect, regular and blocked for reinforced. These are copies of the melee impact particles. Yellow means damage is done, orange means damage is blocked. Does that make sense? IDK but it's how our melee attacks communicate.

I tried impact sounds but it felt a bit much with rapid fire weapons and also a pain to find acceptable sounds. I don't want to increase scope on sounds .

**Review by frawhst — APPROVED** [review]
👍

**Comment by frawhst** [comment]
I forsee people using sec/engi/whatever locked crates to hide things, so we'll probably end up having to apply a similar change to crates ~~and also refactor storage pathing~~ but this looks good for now

**Comment by zay25325** [comment]
this makes it a lot harder to abuse lockers for things like the azone. But that is probably for the best, as you can no longer kill things behind the safety of a locker

**Pull request merged**